lms/views/decorators/h_api.py

```diff
@@ -0,0 +1,126 @@
+"""View decorators for working with the h API."""
+
+import functools
+
+from pyramid.httpexceptions import HTTPBadRequest
+
+from lms.services import HAPINotFoundError
+
+
+__all__ = ["upsert_h_user", "upsert_course_group", "add_user_to_group"]
+
+
+def upsert_h_user(wrapped):
+    """
+    Create or update the Hypothesis user for the request's LTI user.
+
+    Update the h user's information from LTI data. If the user doesn't exist
+    yet, call the h API to create one.
+
+    Assumes that it's only used on LTI launch views.
+    """
+
+    @functools.wraps(wrapped)
+    def wrapper(context, request):
+        if not context.provisioning_enabled:
+            return wrapped(context, request)
+
+        hapi_svc = request.find_service(name="hapi")
+
+        user_data = {
+            "username": context.h_username,
+            "display_name": context.h_display_name,
+            "authority": request.registry.settings["h_authority"],
+            "identities": [
+                {
+                    "provider": context.h_provider,
+                    "provider_unique_id": context.h_provider_unique_id,
+                }
+            ],
+        }
+
+        try:
+            hapi_svc.patch(
+                f"users/{context.h_username}", {"display_name": context.h_display_name}
+            )
+        except HAPINotFoundError:
```

Comment:
Are there any other HTTPish errors that could be encountered here in any sort of expected flow? My hunch is that you're covering these quite well at higher layers but just thought I'd check.

Reply:
This deliberately lets through any `HAPIError`s (the only type of error that `hapi_svc` should raise) other than `HAPINotFoundError`. Since `HAPIError`s are `HTTPInternalServerError`s with appropriate error messages, they end up showing "Connecting to Hypothesis failed" error pages to the user, and reporting the details to Sentry and Papertrail